### ISSUE TITLE
TypeError: Network request failed when fetching a file:// uri

### ISSUE THREAD

**Issue by 0xori** [expo/expo#2402]
### Environment

```
Environment:
  OS: Windows 10
  Node: 8.11.0
  Yarn: 1.7.0
  npm: 5.6.0
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Version  3.1.0.0 AI-173.4907809

Packages: (wanted => installed)
  expo: ^30.0.0 => 30.0.1
  react: 16.3.1 => 16.3.1
  react-native: https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz => 0.55.4
```
App target - Standalone & Client for testing

### Steps to Reproduce

1. use[ this expo example](https://github.com/expo/audio-recording-example) of recording audio
2. mix it with [this example](https://github.com/expo/firebase-storage-upload-example) of uploading to firebase

### Expected Behavior

fetching a file URI from the device file system should return a valid [response](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Response_object).

### Actual Behavior

I'm getting a network error exception 
> TypeError: Network request failed

and if I'm running with remote JS debug mode I can see this error in chrome console
> Failed to load file:///data/user/0/host.exp.exponent/cache/ExperienceData/…/Audio/recording-eee85da3-7121-4273-86a7-686098180662.m4a: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https.

### Some Code
```
const asyncUpload = async ()=> {
          try {   
                const response = await fetch(recordInfo.uri);
                console.log("response", response);
                const blob = await response.blob();
                const blob = new Blob();
                upload( blob)
            }catch (e) {
                console.log("error fetching",e);
            }
}
```

**Comment by nebmesemaou** [comment]
you can use the [FileSystem ](https://docs.expo.io/versions/v30.0.0/sdk/filesystem)api to download remote file or read local file

**Comment by 0xori** [comment]
@nebmesemaou I'm trying to upload a local file to firebase. Tried to use the FileSystem API, I can read the file as a string and then try to upload it, the problem that it's working on android but I'm getting an error for ios. 

> Unhandled promise rejection Error: File 'file:///var/mobile/Containers/Data/Application/7B5B62F5-172E-4D97-884E-DA77052CD847/Documents/ExponentExperienceData/.../myDir/recording-8B438F16-7376-4509-94B7-7BB013F28327.m4a' could not be read.

**Comment by Cheonsoon** [comment]
@0xori file uri should be base64 encode then decode, if you are trying to transfer it through.. As some special char would be url encoded and the FileSystem.read function would use the wrong path.

**Comment by moxorama** [comment]
Also got this issue.

base64 firebase upload also broken on android

**Comment by DhirajKate** [comment]
I am also getting same issue

**Comment by 0xori** [comment]
It happened because I was running it on remote js debug mode, when I turned it off it works

**Comment by mikenz1000** [comment]
> It happened because I was running it on remote js debug mode, when I turned it off it works

That didn't fix it for me.  I'm uploading videos to vimeo so need the blob/buffer/byte array.  readAsStringAsync crashes Expo, I assume because the file is too big (4MB in this case)

**Comment by sjchmiela** [comment]
Great news, @0xori! I'll let myself close this issue then, thank you for updating the thread with your solution! 🙂

**Comment by sjchmiela** [comment]
**TL;DR:** Replace `fetch().blob()` with:
```js
  const blob = await new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.onload = function() {
      resolve(xhr.response);
    };
    xhr.onerror = function() {
      reject(new TypeError('Network request failed'));
    };
    xhr.responseType = 'blob';
    xhr.open('GET', uri, true);
    xhr.send(null);
  });
```

---
Let me guide you through some code to understand the problem better:

Let's start with `fetch` function. It's currently not supported natively by JSC on which JS code runs. When we use React Native, it's being polyfilled by the framework.

Pre-0.57, React Native used a regular `whatwg-fetch` polyfill, implementing `fetch` with `XMLHttpRequest` (`XMLHttpRequest` polyfill has always been implemented by RN), to enable developers to use `fetch` ([usage](https://github.com/facebook/react-native/blob/0.57-stable/Libraries/Network/fetch.js#L14), [dependency definition](https://github.com/facebook/react-native/blob/0.55-stable/package.json#L202)). (`whatwg-fetch` [has set](https://github.com/github/fetch/blob/b59d527c74c7c25078d6cd324a359c84d6857d1e/fetch.js#L455) `xhr.responseType` by default to `"blob"`.) React Native's `XMLHttpRequest` implementation, in turn, [uses](https://github.com/facebook/react-native/blob/099410b8f670d0115ab92d8e064153f8ff04aa04/Libraries/Network/XMLHttpRequest.js#L516) `NetworkingModule` to issue the request. So, at the end, `NetworkingModule.sendRequest` is called.

```
fetch --(whatwg-fetch)--> XMLHttpRequest -> RCTNetworking --(android)--> NetworkingModule.sendRequest
```

![We need to go deeper](https://www.wykop.pl/cdn/c3201142/comment_Edqxb7RnOwTgThjUvf6yatPuGBAW310t.jpg)

Inspecting `NetworkingModule.sendRequest` method we can see that before any real network request is constructed, it [first iterates over `mUriHandlers`](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L254-L261). It's a list of objects conforming to [`UriHandler` interface](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L57-L71), which is very simple, and as the documentation says:

> Allows to implement a custom fetching process for specific URIs. It is the handler's job to fetch the URI and return the JS body payload.

So if we would like to support some custom URI scheme, we would add our own `UriHandler` to `NetworkingModule`, nice! 👌 

In a basic React Native application (on Android), there is only one such custom handler, created by `BlobModule` — [link](https://github.com/facebook/react-native/blob/4148976a83ab96029de1b26b515ff95d3cb58c10/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L77-L103). As we can see from the implementation, it runs only for requests that both:
1. aren't `http://` or `https://` and
2. expect `blob` as response type.

For such requests, this handler is capable of returning a valid blob as expected. So why doesn't it work in RN 0.57+?

[This commit](https://github.com/expo/react-native/commit/122b3791ede095345f44666691aa9ce5aa7f725a), which first landed in React Native 0.57 switched the default `xhr.responseType` from `blob` to `text`. Thus, `BlobModule`'s `UriHandler` doesn't kick in, and since `okhttp` [doesn't recognize URL](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L269) of `file://` scheme, it [throws an exception](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L271).

It's important to note that this commit doesn't introduce a bug. In fact, it fixes a blob leak that was occurring due to `fetch` requests not releasing underlying native blobs when they're done with it!

Since we really need this blob, and can't use `fetch` (there's no way to pass an option that would instruct `fetch` to start with `responseType = blob`) we have to construct `XMLHttpRequest` ourselves.

The following snippet creates a very basic blob promise:
```js
const blob = await new Promise((resolve, reject) => {
   const xhr = new XMLHttpRequest();
   xhr.onload = function() {
     resolve(xhr.response); // when BlobModule finishes reading, resolve with the blob
  };
  xhr.onerror = function() {
    reject(new TypeError('Network request failed')); // error occurred, rejecting
  };
  xhr.responseType = 'blob'; // use BlobModule's UriHandler
  xhr.open('GET', uri, true); // fetch the blob from uri in async mode
  xhr.send(null); // no initial data
});
```

In order not to leak blobs, we should also call `blob.close()` when we're done sending it. That's why the full snippet of `asyncUpload` should look like this:
```js
async function uploadImageAsync(uri) {
  const blob = await new Promise((resolve, reject) => {
    // ... create and resolve with XMLHttpRequest blob
  });

  // do something with the blob, eg. upload it to firebase (API v5.6.0 below)
  const ref = firebase
    .storage()
    .ref()
    .child(uuid.v4());
  const snapshot = await ref.put(blob);
  const remoteUri = await snapshot.ref.getDownloadURL();

  // when we're done sending it, close and release the blob
  blob.close();

  // return the result, eg. remote URI to the image
  return remoteUri;
}
```

**Comment by atrauzzi** [comment]
@sjchmiela - This seems like pretty unpredictable behaviour, can it be considered a bug in react native, despite the fact that doing so prevents a memory problem?

My hope would be that opening a `file:///` uri will still be possible without a ~10 line workaround to trick internal behaviour into doing something that's implied by the API... 😄

**Comment by siddhatiwari** [comment]
Still getting seemingly random 'Network request failed' errors even when using the suggested XMLHttpRequest code snippet.

**Comment by txvnt** [comment]
> 
> 
> **TL;DR:** Replace `fetch().blob()` with:
> 
> ```js
>   const blob = await new Promise((resolve, reject) => {
>     const xhr = new XMLHttpRequest();
>     xhr.onload = function() {
>       resolve(xhr.response);
>     };
>     xhr.onerror = function() {
>       reject(new TypeError('Network request failed'));
>     };
>     xhr.responseType = 'blob';
>     xhr.open('GET', uri, true);
>     xhr.send(null);
>   });
> ```
> 
> Let me guide you through some code to understand the problem better:
> 
> Let's start with `fetch` function. It's currently not supported natively by JSC on which JS code runs. When we use React Native, it's being polyfilled by the framework.
> 
> Pre-0.57, React Native used a regular `whatwg-fetch` polyfill, implementing `fetch` with `XMLHttpRequest` (`XMLHttpRequest` polyfill has always been implemented by RN), to enable developers to use `fetch` ([usage](https://github.com/facebook/react-native/blob/0.57-stable/Libraries/Network/fetch.js#L14), [dependency definition](https://github.com/facebook/react-native/blob/0.55-stable/package.json#L202)). (`whatwg-fetch` [has set](https://github.com/github/fetch/blob/b59d527c74c7c25078d6cd324a359c84d6857d1e/fetch.js#L455) `xhr.responseType` by default to `"blob"`.) React Native's `XMLHttpRequest` implementation, in turn, [uses](https://github.com/facebook/react-native/blob/099410b8f670d0115ab92d8e064153f8ff04aa04/Libraries/Network/XMLHttpRequest.js#L516) `NetworkingModule` to issue the request. So, at the end, `NetworkingModule.sendRequest` is called.
> 
> ```
> fetch --(whatwg-fetch)--> XMLHttpRequest -> RCTNetworking --(android)--> NetworkingModule.sendRequest
> ```
> 
> ![We need to go deeper](https://camo.githubusercontent.com/d457fb4f374bdf9c775824ce93dc00e90db10618/68747470733a2f2f7777772e77796b6f702e706c2f63646e2f63333230313134322f636f6d6d656e745f456471786237526e4f77546754686a55766636796174507547424157333130742e6a7067)
> 
> Inspecting `NetworkingModule.sendRequest` method we can see that before any real network request is constructed, it [first iterates over `mUriHandlers`](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L254-L261). It's a list of objects conforming to [`UriHandler` interface](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L57-L71), which is very simple, and as the documentation says:
> 
> > Allows to implement a custom fetching process for specific URIs. It is the handler's job to fetch the URI and return the JS body payload.
> 
> So if we would like to support some custom URI scheme, we would add our own `UriHandler` to `NetworkingModule`, nice! 👌
> 
> In a basic React Native application (on Android), there is only one such custom handler, created by `BlobModule` — [link](https://github.com/facebook/react-native/blob/4148976a83ab96029de1b26b515ff95d3cb58c10/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L77-L103). As we can see from the implementation, it runs only for requests that both:
> 
>     1. aren't `http://` or `https://` and
> 
>     2. expect `blob` as response type.
> 
> 
> For such requests, this handler is capable of returning a valid blob as expected. So why doesn't it work in RN 0.57+?
> 
> [This commit](https://github.com/expo/react-native/commit/122b3791ede095345f44666691aa9ce5aa7f725a), which first landed in React Native 0.57 switched the default `xhr.responseType` from `blob` to `text`. Thus, `BlobModule`'s `UriHandler` doesn't kick in, and since `okhttp` [doesn't recognize URL](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L269) of `file://` scheme, it [throws an exception](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L271).
> 
> It's important to note that this commit doesn't introduce a bug. In fact, it fixes a blob leak that was occurring due to `fetch` requests not releasing underlying native blobs when they're done with it!
> 
> Since we really need this blob, and can't use `fetch` (there's no way to pass an option that would instruct `fetch` to start with `responseType = blob`) we have to construct `XMLHttpRequest` ourselves.
> 
> The following snippet creates a very basic blob promise:
> 
> ```js
> const blob = await new Promise((resolve, reject) => {
>    const xhr = new XMLHttpRequest();
>    xhr.onload = function() {
>      resolve(xhr.response); // when BlobModule finishes reading, resolve with the blob
>   };
>   xhr.onerror = function() {
>     reject(new TypeError('Network request failed')); // error occurred, rejecting
>   };
>   xhr.responseType = 'blob'; // use BlobModule's UriHandler
>   xhr.open('GET', uri, true); // fetch the blob from uri in async mode
>   xhr.send(null); // no initial data
> });
> ```
> 
> In order not to leak blobs, we should also call `blob.close()` when we're done sending it. That's why the full snippet of `asyncUpload` should look like this:
> 
> ```js
> async function uploadImageAsync(uri) {
>   const blob = await new Promise((resolve, reject) => {
>     // ... create and resolve with XMLHttpRequest blob
>   });
> 
>   // do something with the blob, eg. upload it to firebase (API v5.6.0 below)
>   const ref = firebase
>     .storage()
>     .ref()
>     .child(uuid.v4());
>   const snapshot = await ref.put(blob);
>   const remoteUri = await snapshot.ref.getDownloadURL();
> 
>   // when we're done sending it, close and release the blob
>   blob.close();
> 
>   // return the result, eg. remote URI to the image
>   return remoteUri;
> }
> ```

This worked just fine. Thanks!

**Comment by SteveTro** [comment]
> **TL;DR:** Replace `fetch().blob()` with:
> 
> ```js
>   const blob = await new Promise((resolve, reject) => {
>     const xhr = new XMLHttpRequest();
>     xhr.onload = function() {
>       resolve(xhr.response);
>     };
>     xhr.onerror = function() {
>       reject(new TypeError('Network request failed'));
>     };
>     xhr.responseType = 'blob';
>     xhr.open('GET', uri, true);
>     xhr.send(null);
>   });
> ```
> 
> Let me guide you through some code to understand the problem better:
> 
> Let's start with `fetch` function. It's currently not supported natively by JSC on which JS code runs. When we use React Native, it's being polyfilled by the framework.
> 
> Pre-0.57, React Native used a regular `whatwg-fetch` polyfill, implementing `fetch` with `XMLHttpRequest` (`XMLHttpRequest` polyfill has always been implemented by RN), to enable developers to use `fetch` ([usage](https://github.com/facebook/react-native/blob/0.57-stable/Libraries/Network/fetch.js#L14), [dependency definition](https://github.com/facebook/react-native/blob/0.55-stable/package.json#L202)). (`whatwg-fetch` [has set](https://github.com/github/fetch/blob/b59d527c74c7c25078d6cd324a359c84d6857d1e/fetch.js#L455) `xhr.responseType` by default to `"blob"`.) React Native's `XMLHttpRequest` implementation, in turn, [uses](https://github.com/facebook/react-native/blob/099410b8f670d0115ab92d8e064153f8ff04aa04/Libraries/Network/XMLHttpRequest.js#L516) `NetworkingModule` to issue the request. So, at the end, `NetworkingModule.sendRequest` is called.
> 
> ```
> fetch --(whatwg-fetch)--> XMLHttpRequest -> RCTNetworking --(android)--> NetworkingModule.sendRequest
> ```
> ![We need to go deeper](https://camo.githubusercontent.com/d457fb4f374bdf9c775824ce93dc00e90db10618/68747470733a2f2f7777772e77796b6f702e706c2f63646e2f63333230313134322f636f6d6d656e745f456471786237526e4f77546754686a55766636796174507547424157333130742e6a7067)
> 
> Inspecting `NetworkingModule.sendRequest` method we can see that before any real network request is constructed, it [first iterates over `mUriHandlers`](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L254-L261). It's a list of objects conforming to [`UriHandler` interface](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L57-L71), which is very simple, and as the documentation says:
> 
> > Allows to implement a custom fetching process for specific URIs. It is the handler's job to fetch the URI and return the JS body payload.
> 
> So if we would like to support some custom URI scheme, we would add our own `UriHandler` to `NetworkingModule`, nice! 👌
> 
> In a basic React Native application (on Android), there is only one such custom handler, created by `BlobModule` — [link](https://github.com/facebook/react-native/blob/4148976a83ab96029de1b26b515ff95d3cb58c10/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L77-L103). As we can see from the implementation, it runs only for requests that both:
> 
> 1. aren't `http://` or `https://` and
> 2. expect `blob` as response type.
> 
> For such requests, this handler is capable of returning a valid blob as expected. So why doesn't it work in RN 0.57+?
> 
> [This commit](https://github.com/expo/react-native/commit/122b3791ede095345f44666691aa9ce5aa7f725a), which first landed in React Native 0.57 switched the default `xhr.responseType` from `blob` to `text`. Thus, `BlobModule`'s `UriHandler` doesn't kick in, and since `okhttp` [doesn't recognize URL](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L269) of `file://` scheme, it [throws an exception](https://github.com/facebook/react-native/blob/2151c30474518858a5368013f5ea024588b279b9/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L271).
> 
> It's important to note that this commit doesn't introduce a bug. In fact, it fixes a blob leak that was occurring due to `fetch` requests not releasing underlying native blobs when they're done with it!
> 
> Since we really need this blob, and can't use `fetch` (there's no way to pass an option that would instruct `fetch` to start with `responseType = blob`) we have to construct `XMLHttpRequest` ourselves.
> 
> The following snippet creates a very basic blob promise:
> 
> ```js
> const blob = await new Promise((resolve, reject) => {
>    const xhr = new XMLHttpRequest();
>    xhr.onload = function() {
>      resolve(xhr.response); // when BlobModule finishes reading, resolve with the blob
>   };
>   xhr.onerror = function() {
>     reject(new TypeError('Network request failed')); // error occurred, rejecting
>   };
>   xhr.responseType = 'blob'; // use BlobModule's UriHandler
>   xhr.open('GET', uri, true); // fetch the blob from uri in async mode
>   xhr.send(null); // no initial data
> });
> ```
> 
> In order not to leak blobs, we should also call `blob.close()` when we're done sending it. That's why the full snippet of `asyncUpload` should look like this:
> 
> ```js
> async function uploadImageAsync(uri) {
>   const blob = await new Promise((resolve, reject) => {
>     // ... create and resolve with XMLHttpRequest blob
>   });
> 
>   // do something with the blob, eg. upload it to firebase (API v5.6.0 below)
>   const ref = firebase
>     .storage()
>     .ref()
>     .child(uuid.v4());
>   const snapshot = await ref.put(blob);
>   const remoteUri = await snapshot.ref.getDownloadURL();
> 
>   // when we're done sending it, close and release the blob
>   blob.close();
> 
>   // return the result, eg. remote URI to the image
>   return remoteUri;
> }
> ```

This is working for me on Android. With iOS its throwing an error.
Any idea ?

**Comment by atrauzzi** [comment]
Blobs in RN need to be brought into compliance with JS specs with proper native implementations. I'm not convinced they should even be called Blob currently.

**Comment by Oakyden** [comment]
> **TL;DR:** Replace `fetch().blob()` with:
> 
> ```js
>   const blob = await new Promise((resolve, reject) => {
>     const xhr = new XMLHttpRequest();
>     xhr.onload = function() {
>       resolve(xhr.response);
>     };
>     xhr.onerror = function() {
>       reject(new TypeError('Network request failed'));
>     };
>     xhr.responseType = 'blob';
>     xhr.open('GET', uri, true);
>     xhr.send(null);
>   });
> ```
> 
> Let me guide you through some code to understand the problem better:
> }
> ```


God Damn Hero
